x-pack/auditbeat/module/system/packages/packages.go

```diff
@@ -0,0 +1,356 @@
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Experimental("The %v/%v dataset is experimental", moduleName, metricsetName)
+
+	config := config.NewDefaultConfig()
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, errors.Wrapf(err, "failed to unpack the %v/%v config", moduleName, metricsetName)
+	}
+
+	ms := &MetricSet{
+		BaseMetricSet: base,
+		config:        config,
+		log:           logp.NewLogger(moduleName),
+	}
+
+	if config.ReportChanges {
+		ms.cache = cache.New()
+	}
+
+	return ms, nil
+}
+
+// Fetch collects data about the host. It is invoked periodically.
+func (ms *MetricSet) Fetch(report mb.ReporterV2) {
+	packages, err := getPackages()
+	if err != nil {
+		ms.log.Error(err)
+		report.Error(err)
+	}
+	if packages == nil {
+		return
+	}
+
+	if ms.cache != nil && !ms.cache.IsEmpty() {
+		installed, removed := ms.cache.DiffAndUpdateCache(packages)
+
+		for _, pkgInfo := range installed {
+			report.Event(mb.Event{
+				MetricSetFields: common.MapStr{
+					"status":   "installed",
+					"packages": pkgInfo.(Package).toMapStr(),
+				},
+			})
+		}
+
+		for _, pkgInfo := range removed {
+			report.Event(mb.Event{
+				MetricSetFields: common.MapStr{
+					"status":   "removed",
+					"packages": pkgInfo.(Package).toMapStr(),
+				},
+			})
+		}
+	} else {
+		// Report all installed packages
+		var pkgInfos []common.MapStr
+
+		for _, pkgInfo := range packages {
+			pkgInfos = append(pkgInfos, pkgInfo.(Package).toMapStr())
+		}
+
+		report.Event(mb.Event{
+			MetricSetFields: common.MapStr{
+				"packages": pkgInfos,
+			},
+		})
+
+		if ms.cache != nil {
+			// This will initialize the cache with the current packages
+			ms.cache.DiffAndUpdateCache(packages)
+		}
+	}
+}
+
+func getPackages() ([]cache.Cacheable, error) {
+	host, err := sysinfo.Host()
+	if err != nil {
+		return nil, errors.Wrap(err, "Error getting the OS")
+	}
+
+	hostInfo := host.Info()
+	if hostInfo.OS == nil {
+		return nil, errors.New("No host info")
+	}
+
+	var packages []cache.Cacheable
+
+	switch hostInfo.OS.Family {
+	case "redhat":
+		packages, err = listRPMPackages()
+		if err != nil {
+			err = errors.Wrap(err, "Error getting RPM packages")
+		}
+	case "debian":
+		packages, err = listDebPackages()
+		if err != nil {
+			err = errors.Wrap(err, "Error getting DEB packages")
+		}
+	case "darwin":
+		packages, err = listBrewPackages()
+		if err != nil {
+			err = errors.Wrap(err, "Error getting Homebrew packages")
+		}
+	default:
+		return nil, fmt.Errorf("No logic for getting packages for OS family %v", hostInfo.OS.Family)
+	}
+
+	return packages, err
+}
+
+/*
+The following functions copied from https://github.com/tsg/listpackages/blob/master/main.go
+*/
+func listRPMPackages() ([]cache.Cacheable, error) {
+	format := "%{NAME}|%{VERSION}|%{RELEASE}|%{ARCH}|%{LICENSE}|%{INSTALLTIME}|%{SIZE}|%{URL}|%{SUMMARY}\\n"
+	out, err := exec.Command("/usr/bin/rpm", "--qf", format, "-qa").Output()
```

Comment:
Right, true. Do you think it makes sense to allow Auditbeat to execute `/usr/bin/rpm`, or change to reading the RPM db files in `/var/lib/rpm/*` with the help of a library instead?

Reply:
If we could directly read the database file or use librpm's [database api](http://ftp.rpm.org/api/4.14.0/group__rpmdb.html) and statically link it with Auditbeat that would be the ideal solution in my opinion because we don't have to modify Auditbeat's security posture.

Reply:
I agree, this can it can be handled at a later date. For now you could inject config to disable seccomp for the system test case on Linux that executes `rpm`.